File: src/runtime/DrawingCanvasHitTest.js
.pragma library

// Legacy parity reference for the C++ drawing_canvas_core runtime.
// App QML should call drawingCanvasRuntime instead of importing this file.

function asArray(value) {
    if (!value) {
        return []
    }
    if (Array.isArray(value)) {
        return value
    }
    if (typeof value.length === "number") {
        var result = []
        for (var index = 0; index < value.length; ++index) {
            result.push(value[index])
        }
        return result
    }
    return []
}

function distanceToSegment(px, py, ax, ay, bx, by) {
    var dx = bx - ax
    var dy = by - ay
    var lengthSq = dx * dx + dy * dy
    if (lengthSq <= 0.0000001) {
        var pointDx = px - ax
        var pointDy = py - ay
        return Math.sqrt(pointDx * pointDx + pointDy * pointDy)
    }
    var t = Math.max(0, Math.min(1, ((px - ax) * dx + (py - ay) * dy) / lengthSq))
    var hitX = ax + t * dx
    var hitY = ay + t * dy
    var hitDx = px - hitX
    var hitDy = py - hitY
    return Math.sqrt(hitDx * hitDx + hitDy * hitDy)
}

function pointInPolygon(px, py, points) {
    var inside = false
    var list = asArray(points)
    for (var index = 0, prev = list.length - 1; index < list.length; prev = index++) {
        var current = list[index] || [0, 0]
        var previous = list[prev] || [0, 0]
        var x1 = Number(current[0] || 0)
        var y1 = Number(current[1] || 0)
        var x2 = Number(previous[0] || 0)
        var y2 = Number(previous[1] || 0)
        if (((y1 > py) !== (y2 > py)) && (px < (x2 - x1) * (py - y1) / (y2 - y1) + x1)) {
            inside = !inside
        }
    }
    return inside
}

function angleInArc(angleDeg, startDeg, endDeg) {
    var angle = ((angleDeg % 360) + 360) % 360
    var start = ((Number(startDeg || 0) % 360) + 360) % 360
    var end = ((Number(endDeg || 0) % 360) + 360) % 360
    if (start <= end) {
        return angle >= start && angle <= end
    }
    return angle >= start || angle <= end
}

function unrotatePoint(px, py, cx, cy, rotationDeg) {
    var angle = -Number(rotationDeg || 0) * Math.PI / 180
    var dx = px - cx
    var dy = py - cy
    var cosA = Math.cos(angle)
    var sinA = Math.sin(angle)
    return {
        x: cx + dx * cosA - dy * sinA,
        y: cy + dx * sinA + dy * cosA
    }
}

function objectHitScore(object, x, y, tolerance) {
    var kind = String(object.kind || "")
    if (kind === "point" || kind === "tone_probe") {
        var pointDx = Number(object.x || 0) - x
        var pointDy = Number(object.y || 0) - y
        return Math.sqrt(pointDx * pointDx + pointDy * pointDy)
    }
    if (kind === "line" || kind === "glyph_baseline") {
        return distanceToSegment(x, y, Number(object.x1 || 0), Number(object.y1 || 0), Number(object.x2 || 0), Number(object.y2 || 0))
    }
    if (kind === "polyline") {
        var polyline = asArray(object.points)
        var bestLine = 999
        for (var lineIndex = 1; lineIndex < polyline.length; ++lineIndex) {
            var a = polyline[lineIndex - 1] || [0, 0]
            var b = polyline[lineIndex] || [0, 0]
            bestLine = Math.min(bestLine, distanceToSegment(x, y, Number(a[0] || 0), Number(a[1] || 0), Number(b[0] || 0), Number(b[1] || 0)))
        }
        return bestLine
    }
    if (kind === "circle" || kind === "arc") {
        var circleDx = x - Number(object.cx || 0)
        var circleDy = y - Number(object.cy || 0)
        var radiusDelta = Math.abs(Math.sqrt(circleDx * circleDx + circleDy * circleDy) - Number(object.radius || 0))
        if (kind === "arc") {
            var angle = Math.atan2(circleDy, circleDx) * 180 / Math.PI
            if (!angleInArc(angle, object.start_angle_deg, object.end_angle_deg)) {
                return 999
            }
        }
        return radiusDelta
    }
    if (kind === "rectangle" || kind === "image_reference_frame" || kind === "ascii_crop_frame" || kind === "ascii_cell_region") {
        var left = Number(object.x || 0)
        var top = Number(object.y || 0)
        var right = left + Number(object.width || 0)
        var bottom = top + Number(object.height || 0)
        var centerX = (left + right) / 2
        var centerY = (top + bottom) / 2
        var hitPoint = unrotatePoint(x, y, centerX, centerY, object.rotation_deg)
        var localX = hitPoint.x
        var localY = hitPoint.y
        if (localX >= left && localX <= right && localY >= top && localY <= bottom) {
            return 0
        }
        var clampedX = Math.max(left, Math.min(right, localX))
        var clampedY = Math.max(top, Math.min(bottom, localY))
        var rectDx = localX - clampedX
        var rectDy = localY - clampedY
        return Math.sqrt(rectDx * rectDx + rectDy * rectDy)
    }
    if (kind === "polygon") {
        var polygon = asArray(object.points)
        if (pointInPolygon(x, y, polygon)) {
            return 0
        }
        var bestEdge = 999
        for (var polyIndex = 0; polyIndex < polygon.length; ++polyIndex) {
            var p1 = polygon[polyIndex] || [0, 0]
            var p2 = polygon[(polyIndex + 1) % polygon.length] || [0, 0]
            bestEdge = Math.min(bestEdge, distanceToSegment(x, y, Number(p1[0] || 0), Number(p1[1] || 0), Number(p2[0] || 0), Number(p2[1] || 0)))
        }
        return bestEdge
    }
    return 999
}

function hitObjectAt(objects, x, y, tolerance) {
    var list = asArray(objects)
    var hitTolerance = Math.max(0, Number(tolerance || 0))
    var best = {
        objectId: "",
        kind: "none",
        distance: hitTolerance
    }
    for (var index = list.length - 1; index >= 0; --index) {
        var object = list[index] || ({})
        var score = objectHitScore(object, Number(x), Number(y), hitTolerance)
        if (score < best.distance) {
            best = {
                objectId: String(object.id || ""),
                kind: "object",
                distance: score
            }
        }
    }
    return best
}

File: src/canvas/DrawingCanvasHitTest.cpp
#include "DrawingCanvasHitTest.h"
#include "DrawingCanvasHandles.h"

#include <algorithm>
#include <cmath>

namespace drawing_canvas {
namespace {

constexpr double pi = 3.14159265358979323846;

double distanceToSegment(double px, double py, double ax, double ay, double bx, double by) {
    const double dx = bx - ax;
    const double dy = by - ay;
    if (std::abs(dx) < 0.000001 && std::abs(dy) < 0.000001) {
        const double ox = px - ax;
        const double oy = py - ay;
        return std::sqrt(ox * ox + oy * oy);
    }
    const double t = std::max(0.0, std::min(1.0, ((px - ax) * dx + (py - ay) * dy) / (dx * dx + dy * dy)));
    const double x = ax + t * dx;
    const double y = ay + t * dy;
    const double ox = px - x;
    const double oy = py - y;
    return std::sqrt(ox * ox + oy * oy);
}

bool pointInPolygon(const std::vector<CanvasPoint> &points, double x, double y) {
    if (points.size() < 3) {
        return false;
    }
    bool inside = false;
    for (std::size_t i = 0, j = points.size() - 1; i < points.size(); j = i++) {
        const double xi = finiteNumber(points[i].x, 0.0);
        const double yi = finiteNumber(points[i].y, 0.0);
        const double xj = finiteNumber(points[j].x, 0.0);
        const double yj = finiteNumber(points[j].y, 0.0);
        const bool intersects = ((yi > y) != (yj > y))
            && (x < (xj - xi) * (y - yi) / std::max(0.000001, yj - yi) + xi);
        if (intersects) {
            inside = !inside;
        }
    }
    return inside;
}

double angleDegrees(double x, double y) {
    return std::fmod(std::atan2(y, x) * 180.0 / pi + 360.0, 360.0);
}

bool angleInArc(double angle, double start, double end) {
    const double normalizedAngle = std::fmod(std::fmod(angle, 360.0) + 360.0, 360.0);
    const double normalizedStart = std::fmod(std::fmod(start, 360.0) + 360.0, 360.0);
    const double normalizedEnd = std::fmod(std::fmod(end, 360.0) + 360.0, 360.0);
    if (normalizedStart <= normalizedEnd) {
        return normalizedAngle >= normalizedStart && normalizedAngle <= normalizedEnd;
    }
    return normalizedAngle >= normalizedStart || normalizedAngle <= normalizedEnd;
}

CanvasPoint unrotatePoint(const CanvasObjectView &object, double x, double y) {
    return unrotatePointForRect(object, x, y);
}

} // namespace

double objectHitScore(const CanvasObjectView &object, double x, double y) {
    const QString kind = object.kind();
    if (object.visible() == false) {
        return 999.0;
    }
    if (kind == QStringLiteral("point") || kind == QStringLiteral("tone_probe")) {
        const double dx = x - object.number(QStringLiteral("x"));
        const double dy = y - object.number(QStringLiteral("y"));
        return std::sqrt(dx * dx + dy * dy);
    }
    if (kind == QStringLiteral("line") || kind == QStringLiteral("glyph_baseline")) {
        return distanceToSegment(
            x, y,
            object.number(QStringLiteral("x1")),
            object.number(QStringLiteral("y1")),
            object.number(QStringLiteral("x2")),
            object.number(QStringLiteral("y2")));
    }
    if (kind == QStringLiteral("polyline")) {
        const std::vector<CanvasPoint> points = object.points();
        if (points.size() < 2) {
            return 999.0;
        }
        double best = 999.0;
        for (std::size_t i = 1; i < points.size(); ++i) {
            best = std::min(best, distanceToSegment(x, y, points[i - 1].x, points[i - 1].y, points[i].x, points[i].y));
        }
        return best;
    }
    if (kind == QStringLiteral("circle") || kind == QStringLiteral("arc")) {
        const double cx = object.number(QStringLiteral("cx"));
        const double cy = object.number(QStringLiteral("cy"));
        const double radius = std::max(0.0, object.number(QStringLiteral("radius")));
        const double dx = x - cx;
        const double dy = y - cy;
        const double distance = std::sqrt(dx * dx + dy * dy);
        if (kind == QStringLiteral("arc")
            && !angleInArc(angleDegrees(dx, dy), object.number(QStringLiteral("start_angle_deg")), object.number(QStringLiteral("end_angle_deg")))) {
            return 999.0;
        }
        return std::abs(distance - radius);
    }
    if (isRectangleLike(kind)) {
        const CanvasPoint local = unrotatePoint(object, x, y);
        const double left = object.number(QStringLiteral("x"));
        const double top = object.number(QStringLiteral("y"));
        const double width = object.number(QStringLiteral("width"));
        const double height = object.number(QStringLiteral("height"));
        const double right = left + width;
        const double bottom = top + height;
        if (local.x >= left && local.x <= right && local.y >= top && local.y <= bottom) {
            const double edgeDistance = std::min({local.x - left, right - local.x, local.y - top, bottom - local.y});
            return std::max(0.0, edgeDistance);
        }
        const double clampedX = std::max(left, std::min(right, local.x));
        const double clampedY = std::max(top, std::min(bottom, local.y));
        const double dx = local.x - clampedX;
        const double dy = local.y - clampedY;
        return std::sqrt(dx * dx + dy * dy);
    }
    if (kind == QStringLiteral("polygon")) {
        const std::vector<CanvasPoint> points = object.points();
        if (points.size() < 3) {
            return 999.0;
        }
        if (pointInPolygon(points, x, y)) {
            return 0.0;
        }
        double best = 999.0;
        for (std::size_t i = 0; i < points.size(); ++i) {
            const CanvasPoint a = points[i];
            const CanvasPoint b = points[(i + 1) % points.size()];
            best = std::min(best, distanceToSegment(x, y, a.x, a.y, b.x, b.y));
        }
        return best;
    }
    return 999.0;
}

HitResult hitObjectAt(const std::vector<CanvasObjectView> &objects, double x, double y, double tolerance) {
    HitResult best;
    best.distance = std::max(0.0, finiteNumber(tolerance, 0.02));
    for (auto it = objects.rbegin(); it != objects.rend(); ++it) {
        const double score = objectHitScore(*it, x, y);
        if (score < best.distance) {
            best.ok = true;
            best.objectId = it->id();
            best.kind = QStringLiteral("object");
            best.distance = score;
        }
    }
    if (!best.ok) {
        best.distance = 999.0;
    }
    return best;
}

} // namespace drawing_canvas
